--- a/net/mac80211/tx.c	2024-07-27 10:20:39.999504181 +0200
+++ b/net/mac80211/tx.c	2024-08-06 09:41:14.933436978 +0200
@@ -5297,8 +5297,10 @@
 	if (beacon->tail)
 		skb_put_data(skb, beacon->tail, beacon->tail_len);
 
-	if (ieee80211_beacon_protect(skb, local, sdata, link) < 0)
+	if (ieee80211_beacon_protect(skb, local, sdata, link) < 0) {
+		dev_kfree_skb(skb);
 		return NULL;
+	}
 
 	ieee80211_beacon_get_finish(hw, vif, link, offs, beacon, skb,
 				    chanctx_conf, csa_off_base);
