--- a/mt7915/init.c
+++ b/mt7915/init.c
@@ -411,7 +411,9 @@ mt7915_init_wiphy(struct mt7915_phy *phy
 
 	hw->max_tx_fragments = 4;
 
-	if (!phy->dev->dbdc_support)
+	if (phy->dev->dbdc_support && is_mt7915(&dev->mt76))
+		wiphy->txq_memory_limit = 16 << 20; /* 16 MiB */
+	else
 		wiphy->txq_memory_limit = 32 << 20; /* 32 MiB */
 
 	if (phy->mt76->cap.has_2ghz) {
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -416,6 +416,7 @@ mt7996_init_wiphy(struct ieee80211_hw *h
 	ieee80211_hw_set(hw, SUPPORTS_MULTI_BSSID);
 
 	hw->max_tx_fragments = 4;
+	wiphy->txq_memory_limit = 32 << 20; /* 32 MiB */
 
 	if (phy->mt76->cap.has_2ghz) {
 		phy->mt76->sband_2g.sband.ht_cap.cap |=
