--- a/mt7915/mcu.c
+++ b/mt7915/mcu.c
@@ -2718,8 +2718,8 @@ int mt7915_mcu_rdd_background_enable(str
 	int err, region;
 
 	if (!chandef) { /* disable offchain */
-		err = mt76_connac_mcu_rdd_cmd(&dev->mt76, RDD_STOP, MT_RX_SEL2,
-					      0, 0);
+		err = mt76_connac_mcu_rdd_cmd(&dev->mt76, RDD_STOP, 0,
+					      MT_RX_SEL2, 0);
 		if (err)
 			return err;
 
@@ -2745,8 +2745,8 @@ int mt7915_mcu_rdd_background_enable(str
 		break;
 	}
 
-	return mt76_connac_mcu_rdd_cmd(&dev->mt76, RDD_START, MT_RX_SEL2,
-				       0, region);
+	return mt76_connac_mcu_rdd_cmd(&dev->mt76, RDD_START, 0,
+				       MT_RX_SEL2, region);
 }
 
 int mt7915_mcu_set_chan_info(struct mt7915_phy *phy, int cmd)
