commit 5d5100d708483ce75ba744936298a2c2f8fa6e93 (HEAD -> master)
Author: Łukasz M <lukasz1992@example.com>
Date:   Mon Aug 7 17:47:01 2023 +0200

    Remove connac3 support to reduce size

diff --git a/Makefile b/Makefile
index c8aedf6d..46938692 100644
--- a/Makefile
+++ b/Makefile
@@ -31,7 +31,7 @@ mt76x02-lib-y := mt76x02_util.o mt76x02_mac.o mt76x02_mcu.o \
 
 mt76x02-usb-y := mt76x02_usb_mcu.o mt76x02_usb_core.o
 
-mt76-connac-lib-y := mt76_connac_mcu.o mt76_connac_mac.o mt76_connac3_mac.o
+mt76-connac-lib-y := mt76_connac_mcu.o mt76_connac_mac.o
 
 mt792x-lib-y := mt792x_core.o mt792x_mac.o mt792x_trace.o \
                mt792x_debugfs.o mt792x_dma.o
